File: tools/bench/ii2_parse.cpp
// II-2 §4 — 같은 프레임을 세 가지 방법으로 쪼갠다. 복잡도는 셋 다 O(n).
//
//   g++ -std=c++17 -O2 tools/bench/ii2_parse.cpp -o /tmp/pp && /tmp/pp 200000
//
//   stringstream : getline 으로 한 필드씩. 필드마다 string 객체가 새로 만들어진다
//   substr       : find 로 구분자를 찾고 substr 로 잘라낸다. 역시 복사가 필드마다
//   string_view  : 원본을 가리키는 (포인터, 길이) 쌍만 만든다. 복사가 0회
//
// 갈리는 것은 복사 횟수뿐이다. 파서가 느린 이유는 거의 항상 알고리즘이 아니라
// "필드마다 문자열을 새로 만들고 있다" 이다.
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <sstream>
#include <string>
#include <string_view>
#include <vector>
using namespace std;

int main(int argc, char** argv) {
    int fields = (argc > 1) ? atoi(argv[1]) : 200000;
    string line;
    for (int i = 0; i < fields; ++i) {
        if (i) line += '|';
        line += "field";
        line += to_string(i);
    }

    auto t0 = chrono::steady_clock::now();
    long long n1 = 0;
    {
        istringstream is(line);
        string tok;
        while (getline(is, tok, '|')) n1 += (long long)tok.size();
    }
    auto t1 = chrono::steady_clock::now();

    long long n2 = 0;
    {
        size_t pos = 0;
        while (true) {
            size_t nx = line.find('|', pos);
            string tok = line.substr(pos, nx == string::npos ? string::npos : nx - pos);
            n2 += (long long)tok.size();
            if (nx == string::npos) break;
            pos = nx + 1;
        }
    }
    auto t2 = chrono::steady_clock::now();

    long long n3 = 0;
    {
        string_view sv(line);
        size_t pos = 0;
        while (true) {
            size_t nx = sv.find('|', pos);
            string_view tok = sv.substr(pos, nx == string_view::npos ? string_view::npos : nx - pos);
            n3 += (long long)tok.size();
            if (nx == string_view::npos) break;
            pos = nx + 1;
        }
    }
    auto t3 = chrono::steady_clock::now();

    double a = chrono::duration<double>(t1 - t0).count();
    double b = chrono::duration<double>(t2 - t1).count();
    double c = chrono::duration<double>(t3 - t2).count();
    printf("필드 %d개, 줄 길이 %zu바이트 (합 %lld/%lld/%lld)\n", fields, line.size(), n1, n2, n3);
    printf("stringstream  %.4f s\n", a);
    printf("substr        %.4f s  (%.1f배 빠름)\n", b, a / b);
    printf("string_view   %.4f s  (%.1f배 빠름)\n", c, a / c);
    return 0;
}

File: tools/bench/ii2_parse.py
"""II-2 §4 — Python 에서 같은 프레임을 세 가지로 쪼갠다.

    python3.13 tools/bench/ii2_parse.py 200000

    split     : C 로 구현된 한 번의 훑기. 필드마다 str 객체가 생긴다
    수동 루프 : 파이썬 레벨에서 문자를 하나씩 본다. 같은 O(n) 인데 상수가 다르다
    정규식    : 컴파일된 패턴으로 훑는다

교훈은 C++ 쪽과 반대 방향이다. C++ 에서는 표준 스트림이 느려서 직접 짜는 편이
빠르지만, Python 에서는 표준 함수가 C 구현이라 직접 짜는 순간 느려진다.
"""

import re
import sys
import time

fields = int(sys.argv[1]) if len(sys.argv) > 1 else 200000
line = "|".join(f"field{i}" for i in range(fields))

t0 = time.perf_counter()
n1 = sum(len(tok) for tok in line.split("|"))
t1 = time.perf_counter()

n2 = 0
cur = 0
for ch in line:
    if ch == "|":
        n2 += cur
        cur = 0
    else:
        cur += 1
n2 += cur
t2 = time.perf_counter()

pat = re.compile(r"[^|]*")
n3 = sum(len(m.group()) for m in pat.finditer(line) if m.group())
t3 = time.perf_counter()

a, b, c = t1 - t0, t2 - t1, t3 - t2
print(f"필드 {fields}개, 줄 길이 {len(line)}자 (합 {n1}/{n2}/{n3})")
print(f"split      {a:.4f} s")
print(f"수동 루프  {b:.4f} s  ({b / a:.1f}배 느림)")
print(f"정규식     {c:.4f} s  ({c / a:.1f}배 느림)")
